src/eredis_cluster.erl

```diff
@@ -175,63 +175,99 @@ query(Command, PoolKey) ->
 
-        % If the tcp connection is closed (connection timeout), the redis worker
-        % will try to reconnect, thus the connection should be recovered for
-        % the next request. We don't need to refresh the slot mapping in this
-        % case
-        {error, tcp_closed} ->
+        Payload ->
+            Payload
+    end.
+
+handle_transaction_result(Result) ->
+    case Result of
+        % If instance down we wait refresh finish and retry
+        {error, no_connection} ->
+            eredis_cluster_monitor:refresh_mapping(),
             retry;
 
-        % Redis explicitly say our slot mapping is incorrect, we need to refresh
-        % it
-        {error, <<"MOVED ", _/binary>>} ->
-            eredis_cluster_monitor:refresh_mapping(Version),
+        % If any other error we retry
+        {error, Reason} ->
+            log_error(["Redis Cluster Error: ~p", [Reason]]),
             retry;
 
-        Payload ->
-            Payload
+        % Successful transactions and pool_empty error just return
+        Payload -> Payload
     end.
-handle_transaction_result(Result, Version, check_pipeline_result) ->
-    case handle_transaction_result(Result, Version) of
+handle_transaction_result(Result, check_pipeline_result) ->
+    case handle_transaction_result(Result) of
        retry -> retry;
        Payload when is_list(Payload) ->
            Pred = fun({error, <<"MOVED ", _/binary>>}) -> true;
-                    (_) -> false
-                 end,
+                     (_) -> false
+                  end,
            case lists:any(Pred, Payload) of
-               false -> Payload;
                true ->
-                   eredis_cluster_monitor:refresh_mapping(Version),
-                   retry
+                   eredis_cluster_monitor:refresh_mapping(),
+                   retry;
+               false ->
+                   Payload
            end;
        Payload -> Payload
     end.
 
--spec throttle_retries(integer()) -> ok.
-throttle_retries(0) -> ok;
-throttle_retries(_) -> timer:sleep(?REDIS_RETRY_DELAY).
+try_redirect(SlotAddressPort, Transaction, RedirectionType) ->
+    [Address, Port] = get_address_port(SlotAddressPort),
+    case eredis_cluster_pool:create(Address, Port) of
+        {ok, Pool} -> redirect_transaction(Pool, Transaction, RedirectionType);
+        _ -> retry
+    end.
+
+get_address_port(SlotAddressPort) ->
+    [_Slot, AddressPort | _] = string:split(binary_to_list(SlotAddressPort), " "),
+    [Address, Port] = string:split(AddressPort, ":"),
+    [Address, list_to_integer(Port)].
+
+redirect_transaction(Pool, Transaction, moved) ->
+    eredis_cluster_pool:transaction(Pool, Transaction);
+redirect_transaction(Pool, Transaction, ask) ->
+    % ASK redirection: https://redis.io/topics/cluster-spec
+    Asking = fun(Worker) -> qw(Worker, ["ASKING"]) end,
+    eredis_cluster_pool:transaction(Pool, Asking),
+    eredis_cluster_pool:transaction(Pool, Transaction).
```

Comment:
Hey @zuiderkwast thanks! I will handle 1. Also, great work on the Nordix fork.

Do you have interest in merging this refresh strategy into Nordix? It improves back pressure during resharding.

I would like to use Nordix fork instead of maintaining this repo. Is it stable in production?

Reply:
That's great to hear that you want to collaborate on one repository! I will write you an email.

I'm willing to add your improvements to the Nordix fork, such as exponential backoff and other improvements you have made.

Regarding the refresh strategy, we have done other things such as reusing an existing connection from one of the pools for refresh slots mappings, so it doesn't need to connect. When we get ASK or MOVED, we cast a message to the monitor process to async refresh mapping in the background, then follow the redirect without waiting.

I don't see how the non-blocking monitor calls can improve back pressure. The Version variable solves the same problem. If multiple workers call refresh_mapping, the first one will do the CLUSTER SLOTS query and the others calls will just wait in the inbox of the monitor process until the first one is done and the version is changed. Then, the other calls will be replied immediately, because the Version doesn't match anymore. See these lines:

```Erlang
handle_call({reload_slots_map, Version}, _From, #state{version=Version} = State) ->
    {reply, ok, reload_slots_map(State)};
handle_call({reload_slots_map, _}, _From, State) ->
    %% Mismatching version. Slots map already reloaded.
    {reply, ok, State};
```

What do you think? I might be wrong. Can you explain why your strategy improves back pressure?